# Accuracy calculation using linear regression with reward_function_1

In [51]:
import pandas as pd

df = pd.read_csv('file1.csv')

alpha = 0.90

def rewardget(score):
    if score >= 0 and score <= 19:
        return 50
    elif score >= 20 and score <= 39:
        return 40
    elif score >= 40 and score <= 59:
        return 30
    elif score >= 60 and score <= 79:
        return 20
    elif score >= 80 and score <= 99:
        return 10
    elif score == 100:
        return 0
    elif score >= 101 and score <= 120:
        return -10
    elif score >= 121 and score <= 140:
        return -20
    elif score >= 141 and score <= 160:
        return -30
    elif score >= 161 and score <= 180:
        return -40
    elif score >= 181 and score <= 200:
        return -50

def valuefind(reward, previousValue):
    currentValue = reward + alpha * previousValue
    return currentValue

currentValues = []
previousValue = 0

for i in range(len(df)):
    score = df.loc[i, 'score']
    reward = rewardget(score)
    currentValue = valuefind(reward, previousValue)
    currentValues.append(currentValue)
    previousValue = currentValue

# Append 'currentValues' as a new column to the DataFrame
df['currentValues'] = currentValues

# df.head(20)


In [52]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# seperate input and output features
X = df[['score', 'attack']]
y = df['currentValues']

# train_test_split with declaration of test size
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.15)

# Initialize and train your model
model = LinearRegression()
model.fit(X_train, y_train)

# predictions
y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
accuracy = 1 - (mse / y_test.var())
print("Accuracy:", accuracy)


Accuracy: 0.19789896341494007


# Accuracy calculation using linear regression with reward_function_2

In [54]:
import pandas as pd

df = pd.read_csv('file1.csv')

alpha = 0.90
previousValues = []

def rewardget(score, previousValues):
    previousValues_mean = sum(previousValues[-10:]) / len(previousValues[-10:]) if previousValues else 0
    if score >= 0 and score <= 19:
        reward = 50 - (previousValues_mean / 10)
    elif score >= 20 and score <= 39:
        reward = 40 - (previousValues_mean / 10)
    elif score >= 40 and score <= 59:
        reward = 30 - (previousValues_mean / 10)
    elif score >= 60 and score <= 79:
        reward = 20 - (previousValues_mean / 10)
    elif score >= 80 and score <= 99:
        reward = 10 - (previousValues_mean / 10)
    elif score == 100:
        reward = 0 - (previousValues_mean / 10)
    elif score >= 101 and score <= 120:
        reward = -10 - (previousValues_mean / 10)
    elif score >= 121 and score <= 140:
        reward = -20 - (previousValues_mean / 10)
    elif score >= 141 and score <= 160:
        reward = -30 - (previousValues_mean / 10)
    elif score >= 161 and score <= 180:
        reward = -40 - (previousValues_mean / 10)
    elif score >= 181 and score <= 200:
        reward = -50 - (previousValues_mean / 10)
    return reward

def valuefind(reward, previousValue):
    currentValue = reward + alpha * previousValue
    return currentValue

currentValues = []
previousValue = 0

for i in range(len(df)):
    score = df.loc[i, 'score']
    reward = rewardget(score, previousValues)
    currentValue = valuefind(reward, previousValue)
    currentValues.append(currentValue)
    previousValues.append(currentValue)
    previousValue = currentValue

# Append 'currentValues' as a new column to the DataFrame
df['currentValues'] = currentValues

# df.head(20)

In [55]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# seperate input and out features
X = df[['score', 'attack']]
y = df['currentValues']

# train_test_split with test size declaration
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.15)

model = LinearRegression()
model.fit(X_train, y_train)

# predictions
y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
accuracy = 1 - (mse / y_test.var())
print("Accuracy:", accuracy)

Accuracy: 0.28120987259248253


# Accuracy calculation using linear regression with reward_function_3

In [57]:
import pandas as pd

df = pd.read_csv('file1.csv')

alpha = 0.90
previousValues = []

def rewardget(score, previousValues):
    previousValues_mean = sum(previousValues[-10:]) / len(previousValues[-10:]) if previousValues else 0
    if score >= 0 and score <= 19:
        reward = 50 - (previousValues_mean / 10)
    elif score >= 20 and score <= 39:
        reward = 40 - (previousValues_mean / 10)
    elif score >= 40 and score <= 59:
        reward = 30 - (previousValues_mean / 10)
    elif score >= 60 and score <= 79:
        reward = 20 - (previousValues_mean / 10)
    elif score >= 80 and score <= 99:
        reward = 10 - (previousValues_mean / 10)
    elif score == 100:
        reward = 0 - (previousValues_mean / 10)
    elif score >= 101 and score <= 120:
        reward = -10 - (previousValues_mean / 10)
    elif score >= 121 and score <= 140:
        reward = -20 - (previousValues_mean / 10)
    elif score >= 141 and score <= 160:
        reward = -30 - (previousValues_mean / 10)
    elif score >= 161 and score <= 180:
        reward = -40 - (previousValues_mean / 10)
    elif score >= 181 and score <= 200:
        reward = -50 - (previousValues_mean / 10)
    return reward

def valuefind(reward, previousValue, prev_score, current_score):
    if abs(current_score - prev_score) > 100:
        currentValue = 2 * reward + alpha * previousValue
    else:
        currentValue = reward + alpha * previousValue
    return currentValue

currentValues = []
previousValue = 0
prev_score = df.loc[0, 'score']  # Initialize prev_score with the first score in the dataframe

for i in range(len(df)):
    score = df.loc[i, 'score']
    reward = rewardget(score, previousValues)
    currentValue = valuefind(reward, previousValue, prev_score, score)
    currentValues.append(currentValue)
    previousValues.append(currentValue)
    previousValue = currentValue
    prev_score = score

# Append 'currentValues' as a new column to the DataFrame
df['currentValues'] = currentValues

# df.head(20)

In [61]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# Separate input features and output
X = df[['score', 'attack']]
y = df['currentValues']

# Split the dataframe
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.15)

model = LinearRegression()
model.fit(X_train, y_train)

# predictions
y_pred = model.predict(X_test)

# Calculate accuracy
mse = mean_squared_error(y_test, y_pred)
accuracy = 1 - (mse / y_test.var())
print("Accuracy:", accuracy)

Accuracy: 0.394502225268604


# Accuracy calculation using linear regression with reward_function_4

In [62]:
import pandas as pd

df = pd.read_csv('file1.csv')

alpha = 0.90

import numpy as np

def rewardget(score):
    if score == 100:
        return 0
    else:
        return np.tanh((100-score)/10)*50

def valuefind(reward, previousValue):
    currentValue = reward + alpha * previousValue
    return currentValue

currentValues = []
previousValue = 0

for i in range(len(df)):
    score = df.loc[i, 'score']
    reward = rewardget(score)
    currentValue = valuefind(reward, previousValue)
    currentValues.append(currentValue)
    previousValue = currentValue

df['currentValues'] = currentValues

# df.head(20)

In [63]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

X = df[['score', 'attack']]
y = df['currentValues']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.15)

model = LinearRegression()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
accuracy = 1 - (mse / y_test.var())
print("Accuracy:", accuracy)

Accuracy: 0.1597040027855846


# Accuracy calculation using linear regression with reward_function_5

In [64]:
import pandas as pd
import numpy as np

df = pd.read_csv('file1.csv')

alpha = 0.90
previousValues = []

def rewardget_combined(score, previousValues):
    # reward function 1
    reward_1 = np.tanh((100 - score) / 10) * 50
    
    # reward function 2
    previousValues_mean = sum(previousValues[-10:]) / len(previousValues[-10:]) if previousValues else 0
    if score >= 0 and score <= 19:
        reward_2 = 50 - (previousValues_mean / 10)
    elif score >= 20 and score <= 39:
        reward_2 = 40 - (previousValues_mean / 10)
    elif score >= 40 and score <= 59:
        reward_2 = 30 - (previousValues_mean / 10)
    elif score >= 60 and score <= 79:
        reward_2 = 20 - (previousValues_mean / 10)
    elif score >= 80 and score <= 99:
        reward_2 = 10 - (previousValues_mean / 10)
    elif score == 100:
        reward_2 = 0 - (previousValues_mean / 10)
    elif score >= 101 and score <= 120:
        reward_2 = -10 - (previousValues_mean / 10)
    elif score >= 121 and score <= 140:
        reward_2 = -20 - (previousValues_mean / 10)
    elif score >= 141 and score <= 160:
        reward_2 = -30 - (previousValues_mean / 10)
    elif score >= 161 and score <= 180:
        reward_2 = -40 - (previousValues_mean / 10)
    elif score >= 181 and score <= 200:
        reward_2 = -50 - (previousValues_mean / 10)
    
    # Combine rewards using a weighted average
    reward = alpha * reward_1 + (1 - alpha) * reward_2
    
    return reward

def valuefind(reward, previousValue, prev_score, current_score):
    if abs(current_score - prev_score) > 100:
        currentValue = 2 * reward + alpha * previousValue
    else:
        currentValue = reward + alpha * previousValue
    return currentValue

currentValues = []
previousValue = 0
prev_score = df.loc[0, 'score']

for i in range(len(df)):
    score = df.loc[i, 'score']
    reward = rewardget_combined(score, previousValues)
    currentValue = valuefind(reward, previousValue, prev_score, score)
    currentValues.append(currentValue)
    previousValues.append(currentValue)
    previousValue = currentValue
    prev_score = score

df['currentValues'] = currentValues

# df.head(50)

In [65]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

X = df[['score', 'attack']]
y = df['currentValues']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.15)

model = LinearRegression()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
accuracy = 1 - (mse / y_test.var())
print("Accuracy:", accuracy)

Accuracy: 0.2850929414513794


# Conclusion:
### I got a less accuracy(arround 15-40) the reason for this may be the relation between input and output feature is nonlinear or input features in my dataframe is very less (here only two input features score and attack), here i not used query as a input features because it is string variable and also query column contribution is negligible
### In all the above while i am using reward_function_3 then i got a max accuracy (approximatly 0.39), the reward_function_3 better than other reward_function because in this reward_function I consider 10 previousValues as well as if absolute difference between the current score and just previous score is greater than the 100 then we double the reward it means bellman equation becames currentValue = 2 * reward + alpha * previousValue else currentValue = reward + alpha * previousValue.